src/Symfony/Component/ObjectMapper/CachedObjectMapper.php

```diff
@@ -0,0 +1,351 @@
+ *
+ * @author Antoine Bluchet <soyuka@gmail.com>
+ */
+final class CachedObjectMapper implements ObjectMapperInterface, ObjectMapperAwareInterface
+{
+    use ObjectMapperAwareTrait;
+    use ObjectMapperTrait;
+
+    private array $cachedMappings = [];
+    private ?Filesystem $fs = null;
+
+    /**
+     * Tracks recursive references.
+     */
+    private ?\SplObjectStorage $objectMap = null;
+
+    public function __construct(
+        private readonly string $cacheDir,
+        ObjectMapperMetadataFactoryInterface $metadataFactory = new ReflectionObjectMapperMetadataFactory(),
+        ?PropertyAccessorInterface $propertyAccessor = null,
+        ?ContainerInterface $transformCallableLocator = null,
+        ?ContainerInterface $conditionCallableLocator = null,
+        ?ObjectMapperInterface $objectMapper = null,
+        private readonly bool $debug = false,
+    ) {
+        $this->metadataFactory = $metadataFactory;
+        $this->propertyAccessor = $propertyAccessor;
+        $this->transformCallableLocator = $transformCallableLocator;
+        $this->conditionCallableLocator = $conditionCallableLocator;
+        $this->objectMapper = $objectMapper;
+    }
+
+    public function map(object $source, object|string|null $target = null): object
+    {
+        $objectMapInitialized = false;
+        if (null === $this->objectMap) {
+            $this->objectMap = new \SplObjectStorage();
+            $objectMapInitialized = true;
+        }
+
+        $sourceClass = $source::class;
+        $targetClass = $this->resolveTargetClass($source, $target);
+
+        if (!$targetClass) {
+            throw new MappingException(\sprintf('Mapping target not found for source "%s".', get_debug_type($source)));
+        }
+
+        if (\is_string($target) && !class_exists($target)) {
+            throw new MappingException(\sprintf('Mapping target class "%s" does not exist for source "%s".', $target, get_debug_type($source)));
+        }
+
+        $cacheKey = str_replace(['\\', '/'], '_', $sourceClass.'-to-'.$targetClass);
+
+        if (!isset($this->cachedMappings[$cacheKey])) {
+            $cacheFile = $this->cacheDir.'/'.$cacheKey.'.php';
+
+            if (!file_exists($cacheFile) || $this->debug) {
+                $this->fs ??= new Filesystem();
+                $mappingData = $this->getMappingMetadata($source, $targetClass);
+                $code = $this->generateMappingCode($mappingData, $sourceClass, $targetClass);
+
+                $tmpFile = $this->fs->tempnam(\dirname($cacheFile), basename($cacheFile));
+
+                try {
+                    $this->fs->dumpFile($tmpFile, $code);
+                    $this->fs->rename($tmpFile, $cacheFile, true);
+                    $this->fs->chmod($cacheFile, 0o666 & ~umask());
+                } catch (IOException $e) {
+                    throw new MappingException(\sprintf('Failed to write "%s" mapping file.', $cacheFile), previous: $e);
```

Comment:
all this is already handled by $this->fs->dumpFile($tmpFile, $code);
about the exception, I'd suggest leaving the one from the FS component - a `MappingException` doesn't look correct to me